proposed/swift-interop.md

```diff
@@ -0,0 +1,183 @@
+
+##### Structs/Enums
+
+Like .NET, Swift has both value types and class types. The value types in Swift include both structs and enums. When passed at the ABI layer, they are generally treated as their composite structure and passed by value in registers. However, when Library Evolution mode is enabled, struct layouts are considered "opaque" and their size, alignment, and layout can vary between compile-time and runtime. As a result, all structs and enums in Library Evolution mode are passed by a pointer instead of in registers. Frozen structs and enums (annotated with the `@frozen` attribute) are not considered opaque and will be enregistered. We plan to interoperate with Swift through the Library Evolution mode, so we will generally be able to pass structures using opaque layout.
+
+At the lowest level of the calling convention, we do not consider Library Evolution to be a different calling convention than the Swift calling convention. Library Evolution requires that some types are passed by a pointer/reference, but it does not fundamentally change the calling convention.
+
+### Projecting Swift into .NET
+
+The majority of the work for Swift/.NET interop is determining how a type that exists in Swift should exist in .NET and what shape it should have. This section is a work in progress and will discuss how each feature in Swift will be projected into .NET, particularly in cases where there is not a corresponding .NET or C# language feature. Each feature should have a subheading describing how the projection will look and how any mechanisms to make it work will be designed.
+
+All designs in this section should be designed such that they are trimming and AOT-compatible by construction. We should work to ensure that no feature requires whole-program analysis (such as custom steps in the IL Linker) to be trim or AOT compatible.
+
+#### Swift to .NET Language Feature Projections
+
+##### Structs/Value Types
+
+Unlike .NET, Swift's struct types have strong lifetime semantics more similar to C++ types than .NET structs. At the Swift ABI layer, there are broadly three types of structs/enums: "POD/Trivial" structs, "Bitwise Takable/Movable" structs, and non-bitwise movable structs. The [Swift documentation](https://github.com/apple/swift/blob/main/docs/ABIStabilityManifesto.md#layout-and-properties-of-types) covers these different kinds of structs. Let's look at how we could map each of these categories of structs into .NET.
```

Comment:
> how we could map each of these categories of structs into .NET.

I assume we will need some indication of when a `struct` is "POD/Trivial" or "Bitwise Takable/Movable"? The last one appears to be akin to implementing the type as a `class`, if not will we need another flag/inidication?

Reply:
I believe we can treat "Bitwise Takable/Movable" structs as a set of individual values (which we wrap into a .NET struct) instead of having to treat it as a Swift struct (with Swift lifetime rules) when marshalled into .NET. I believe we could design the projection tooling to not need an indication at the managed code layer for "POD/Trivial" vs "Bitwise Takable/Movable". The non-bitwise-movable structs definitely need to be represented with a `class`.

In any case, the distinction here only matters for `@frozen` structs where their layout can be known at compile time. For all opaque structs, we'll need to treat them as non-bitwise-movable until proven otherwise. Based on @stephen-hawley's comments, trying to determine the difference here may be very expensive and bug-prone, especially at runtime. Having one solution here for all structs may be the best option.